runners/java-fn-execution/src/test/java/org/apache/beam/runners/fnexecution/environment/SingletonDockerEnvironmentManagerTest.java

```diff
@@ -0,0 +1,122 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.runners.fnexecution.environment;
+
+import static com.google.common.truth.Truth.assertThat;
+import static org.junit.Assert.fail;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+import org.apache.beam.model.pipeline.v1.Endpoints.ApiServiceDescriptor;
+import org.apache.beam.model.pipeline.v1.RunnerApi.Environment;
+import org.apache.beam.runners.fnexecution.GrpcFnServer;
+import org.apache.beam.runners.fnexecution.artifact.ArtifactRetrievalService;
+import org.apache.beam.runners.fnexecution.control.SdkHarnessClient;
+import org.apache.beam.runners.fnexecution.control.SdkHarnessClientControlService;
+import org.apache.beam.runners.fnexecution.logging.GrpcLoggingService;
+import org.apache.beam.runners.fnexecution.provisioning.StaticGrpcProvisionService;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.mockito.Mock;
+import org.mockito.Mockito;
+import org.mockito.MockitoAnnotations;
+
+/** Tests for {@link SingletonDockerEnvironmentManager}. */
+@RunWith(JUnit4.class)
+public class SingletonDockerEnvironmentManagerTest {
+
+  private static final ApiServiceDescriptor SERVICE_DESCRIPTOR = ApiServiceDescriptor.newBuilder()
+      .setUrl("service-url")
+      .build();
+  private static final String IMAGE_NAME = "my-image";
+  private static final Environment ENVIRONMENT = Environment.newBuilder()
+      .setUrl(IMAGE_NAME)
+      .build();
+  private static final String CONTAINER_ID =
+      "e4485f0f2b813b63470feacba5fe9cb89699878c095df4124abd320fd5401385";
+
+  @Mock private DockerWrapper docker;
+
+  @Mock private GrpcFnServer<SdkHarnessClientControlService> controlServiceServer;
+  @Mock private GrpcFnServer<GrpcLoggingService> loggingServiceServer;
+  @Mock private GrpcFnServer<ArtifactRetrievalService> retrievalServiceServer;
+  @Mock private GrpcFnServer<StaticGrpcProvisionService> provisioningServiceServer;
+
+  @Mock private SdkHarnessClientControlService clientControlService;
```

Comment:
I believe this is rm'd in HEAD

Reply:
That change hasn't been submitted yet. Is the plan to actually remove it or not?

Reply:
Because the `FnApiControlClientPoolService` uses a blocking queue for the client pool, interacting with it and testing these interactions becomes difficult. I ended up effectively re-implementing this interface in order to mock out the incoming clients. At this point I think it's worth keeping `SdkHarnessClientControlService`.